docs/source/sparse.rst

```diff
@@ -1,146 +1,382 @@
+    >>> s.to_dense()
+    tensor([[0, 0],
+            [0, 0],
+            [1, 3],
+            [0, 0],
+            [5, 7]])
+
+where ``s`` is a (1+1)-dimensional :ref:`hybrid tensor
+<sparse-hybrid-docs>`:
+
+    >>> s[0]
+    tensor([0, 0])
+    >>> s[2]
+    tensor([1, 3])
+
+Notice that slicing in dense dimensions is supported (except with
+negative step):
+
+    >>> s[2, 1:]
+    tensor([3])
+
+In general, a (M + K)-dimensional coalesced sparse COO tensor ``s``
+with the number of specified elements ``nse`` has the following
+invariants:
+
+    >>> s = s.coalesce()
+    >>> M, K = s.sparse_dim(), s.dense_dim()
+    >>> assert M + K == len(s.shape) == s.ndim
+    >>> assert s.indices().shape == (M, nse)
+    >>> assert s.values().shape == (nse,) + s.shape[M:]
+
+Since ``s.indices()`` is always a 2D tensor, the smallest sparse
+dimension M = 1.  Therefore, representation of a sparse tensor of M =
+0 is simply a dense tensor.
+
+.. _sparse-uncoalesed-coo-docs:
+
+Uncoalesced sparse COO tensors
+------------------------------
+
+Torch sparse COO tensor format permits *uncoalesced* sparse tensors,
+where there may be duplicate coordinates in the indices; in this case,
+the interpretation is that the value at that index is the sum of all
+duplicate value entries:
+
+    >>> i = [[1, 1]]
+    >>> v =  [3, 4]
+    >>> s=torch.sparse_coo_tensor(i, v, (3,))
+    >>> s
+    tensor(indices=tensor([[1, 1]]),
+           values=tensor(  [3, 4]),
+           size=(3,), nnz=2, layout=torch.sparse_coo)
+    >>> s.to_dense()
+    tensor([0, 7, 0])
 
-SparseTensor has the following invariants:
-  1. sparse_dim + dense_dim = len(SparseTensor.shape)
-  2. SparseTensor._indices().shape = (sparse_dim, nnz)
-  3. SparseTensor._values().shape = (nnz, SparseTensor.shape[sparse_dim:])
+Uncoalesced tensors permit us to implement certain operators more
+efficiently.  For the most part, you shouldn't have to care whether or
+not a sparse tensor is coalesced or not, as most operations will work
+identically given a coalesced or uncoalesced sparse tensor.  However,
+there are two cases in which you may need to care.
 
-Since SparseTensor._indices() is always a 2D tensor, the smallest sparse_dim = 1.
-Therefore, representation of a SparseTensor of sparse_dim = 0 is simply a dense tensor.
+First, if you repeatedly perform an operation that can produce
+duplicate entries (e.g., :func:`torch.Tensor.add`), you should
+occasionally coalesce your sparse tensors to prevent them from growing
+too large:
+
+    >>> s.add_(torch.sparse_coo_tensor([[2, 2]], [5, 6]))
+    tensor(indices=tensor([[1, 1, 2, 2]]),
+           values=tensor(  [3, 4, 5, 6]),
+           size=(3,), nnz=4, layout=torch.sparse_coo)
+    >>> s.coalesce()
+    tensor(indices=tensor([[1,  2]]),
+           values=tensor(  [7, 11]),
+           size=(3,), nnz=2, layout=torch.sparse_coo)
+
+.. note::
+
+   A coalesced tensor is a sparse COO tensor with unique indices in
```

Comment:
This description should be in the mathematical description of coordinate list tensors. 

Is the expression "sparse COO tensor" redundant? That is, are there "dense" coordinate list tensors? I guess every value in a tensor could be specified using a coordinate list.

Reply:
The notion of "coalesced tensor" makes sense only for sparse COO tensors. For instance, the "uncoalesced" sparse CRS tensor is not a thing. Also, not all sparse tensor formats store indices explicitly.

The expression "sparse COO tensor" is only redundant when one knows what is a coalesced tensor. But here we define what is coalesced tensor in terms of the used storage format and using conditions on the storage format data.

Reply:
Sorry I don't think I understand. Let me know if it'd be more helpful to discuss this directly over Slack or a VC.

I get that the terms coalesced/uncoalesced apply specifically to coordinate list tensors. That's why I was suggesting the concept be included with the description of them. I'm not sure I understand the second paragraph, however. From the language elsewhere in this doc, isn't "sparse" just an imprecise adjective for when the number of specified to unspecified elements is "small"? I guess I'm suggesting it's not so much redundant as not really part of the storage format, which is coordinate list. I'm not sure I understand how the concept of coalescing applies or doesn't apply to the description of the memory format, though?

Reply:
Resolving as obsolete. See the new text.